text/0000-throw-expr.md

```diff
@@ -332,6 +332,23 @@ As always, when a new real keyword is introduced, there is some degree of breaka
 This is a drawback of this proposal. A mitigating factor is that the degree is
 believed to be small.
 
+## *"But what about the success case?"*
+
+If we introduce `throw expr`, then an obvious question becomes:
+\- *"How do you do an early return with Ok-wrapping?"*
+Not being able to perform an early return for the success case could
+be considered a drawback. Although that could also be considered letting
+the perfect be the enemy of the good.
+
+A strategy for introducing early-return on success with Ok-wrapping **could**
+be to introduce `try fn` and / or to let `return` perform an early return
```

Comment:
Or, go the other way around- let people `break` out of `try` blocks. On its own, this is probably a bad idea, as `try` blocks are not loops, but it's possible `'label: try { .. break 'label value; .. }` would be sufficient. For that matter, maybe *any* `return` *or* `break` that leaves the `try` block should be `Ok`-wrapped? For example:

```rust
fn f() -> Result<T, U> {
    try {
        let t: T = get_a_t()?;
        return t;
    }
}
```

Maybe that's even worse.

Reply:
My main concern wrt. re-purposing `break` is this:

```rust
try {
    // ..
    loop {
        // Are we breaking the loop or the try?
        // Perhaps it is obvious to say "the loop", but is it really?
        break 3;
    }
    // ..
}
```

> Maybe that's even worse.

Probably, yeah.

Reply:
Regarding break with label:
```rust
'label: try {
    ..
    break 'label value;
    ..
}
```

That seems workable, but it is not particularly ergonomic.

Reply:
While I'm a fan of _having_ label-break-value, I see it mostly as a way for macros to provide customized flow control constructs.  I'd rather it typically be internal to libraries more than something that shows up in syntax that would be used broadly.

Reply:
Another, possibly bad, idea to facilitate for macros could be to introduce the special label `'try` which always refers to the closest `try { .. }` block.

That way, you could just write:
```rust
try {
    ok!(value);
}
```
where `ok!(value)` expands to `break 'try Try::from_ok(value)`.

Reply:
I'm really not a fan of special labels like `'fn` or `'try`. I would much rather we just stick with `fail` or `throw`, as proposed, than *ever* introduce those.

Reply:
@rpjohnst I'm not a fan either :)

Reply:
I think special labels like `'try` and `'fn` going with `try` and `fn` fits well with `'static` and `static`.

Reply:
@squishy-clouds That's a good point. Tho, `'static` is always usable, but `'try` and `'throw` would not. I don't know if that is a major drawback or even a drawback at all.

Reply:
@Centril I think my rigorous reason is I don't like introducing new names "out of thin air." Every other label has its name chosen by the source code- I would be much less bothered by some kind of `'my_label: fn f() { ... }` or `'my_label: try { ... }` than I am by introducing `'fn` or `'try` as "special" names.

A similar-but-different situation came up in https://github.com/rust-lang/rfcs/pull/2115, where people felt wary about allowing the source code to introduce names without some primary declaration point. But in that case, the source code is still the thing determining the name- you write the same name in two places and they mean the same thing. Labels inherently have a "declaration" point, but the same reasoning applies- *you* write the name down in *both* places, *as a label*.

(@squishy-clouds `'static` is not a label at all- it's a lifetime, and further kind of a "monoidal identity element" like `0` or `1` or `[]`, so it gets a bit of a pass here.)

Reply:
@rpjohnst I don't think they would be so "out of thin air";

If the system of having special labels is applied out consistently per control flow block-form you could have:

+ `'fn`
+ `'loop`
    + `'while`
    + `'for`
    + maybe just merge these as one `loop`?
+ `'if`
+ `'match`
+ `'try`
+ `'async`
+ `'const` (maybe not?)

The goal of these are not to be used directly in people's code, but rather hidden away in macros so that you can invent new and interesting DSLs. The problem with having to explicitly declare the labels is that you have to pass them into the macros.

Reply:
Do we have any actual use cases for a generalized "magic labels" feature like this, even in DSLs? I've never heard anyone suggest any of these except as an implementation detail of a throw/fail macro.

Reply:
One use-case might be that if we had a `'try` built-in label that can be targeted by `?`, I think all of the other exception like syntax can be implemented by the eco-system (or even `std` if deemed worthy) as macros. That would allow for:

* People to use break-with-value and `?` with these blocks without auto-wrapping or any other additional functionality, basically keeping it as close to current control flow as possible.
* Experimentation with more exception like syntax on crates.io. Before things would settle that would also provide an explicit opt-in to specific semantics. Once things have settled, the functionality can simply be added to `std`. Since they would still be macros, there's still an opt-in to additional semantics like auto-converting a final result. Adjusting pieces of `std` also seems easier and more future proof than changing the language itself.

I'm wondering if it would make sense to put together a proposal for this path for the new exception like syntax extensions.

Reply:
> Do we have any actual use cases for a generalized "magic labels" feature like this, even in DSLs?

Not very concretely at least; Just hypothesizing ;)